scripts/src/java/org/oppia/android/scripts/todo/TodoCheck.kt

```diff
@@ -0,0 +1,302 @@
+    checkIfOpenIssueFailure(
+      codeLine = todo.lineContent,
+      openIssueList = openIssueList,
+      correctTodoFormatRegex = correctTodoFormatRegex
+    )
+  }
+
+  val redundantExemptions = retrieveRedundantExemptions(
+    todos = poorlyFormattedTodos + openIssueFailureTodos,
+    todoExemptionList = todoExemptionList,
+    repoPath = repoPath
+  )
+
+  val poorlyFormattedTodosAfterExemption = retrieveTodosAfterExemption(
+    todos = poorlyFormattedTodos,
+    todoExemptionList = todoExemptionList,
+    repoPath = repoPath
+  )
+
+  val openIssueFailureTodosAfterExemption = retrieveTodosAfterExemption(
+    todos = openIssueFailureTodos,
+    todoExemptionList = todoExemptionList,
+    repoPath = repoPath
+  )
+
+  logRedundantExemptions(redundantExemptions, todoExemptionTextProtoFilePath)
+
+  logFailures(
+    invalidTodos = poorlyFormattedTodosAfterExemption,
+    failureMessage = "TODOs not in correct format:",
+    failureNote =
+      "The TODO should be in the format: TODO(#ISSUE_NUMBER): <todo_description>"
+  )
+
+  logFailures(
+    invalidTodos = openIssueFailureTodosAfterExemption,
+    failureMessage = "TODOs not corresponding to an open issue:",
+    failureNote =
+      "Note that, every TODO must correspond to an open issue on GitHub"
+  )
+
+  if (
+    redundantExemptions.isNotEmpty() ||
+    poorlyFormattedTodosAfterExemption.isNotEmpty() ||
+    openIssueFailureTodosAfterExemption.isNotEmpty()
+  ) {
+    throw Exception("TODO CHECK FAILED")
+  } else {
+    println("TODO CHECK PASSED")
+  }
+}
+
+/**
+ * Retrieves the todo failures list after filtering them from the exemptions.
+ *
+ * @param todos the list of all the failure causing todos
+ * @param todoExemptionList the list contating the todo exemptions
+ * @param repoPath path of the repo to be analyzed
+ * @return list obtained after filtering the exemptions
+ */
+private fun retrieveTodosAfterExemption(
+  todos: List<Todo>,
+  todoExemptionList: List<TodoExemption>,
+  repoPath: String
+): List<Todo> {
+  return todos.filter { todo ->
+    !todoExemptionList.any { it ->
+      it.exemptedFilePath == todo.filePath.removePrefix(repoPath) &&
+        todo.lineNumber in it.getLineNumberList()
+    }
+  }
+}
+
+/**
+ * Retrieves a list of redundant exemptions.
+ *
+ * @param todos the list of all the failure causing todos
+ * @param todoExemptionList the list contating the todo exemptions
+ * @param repoPath path of the repo to be analyzed
+ * @return a list of all the redundant exemptions
+ */
+private fun retrieveRedundantExemptions(
+  todos: List<Todo>,
+  todoExemptionList: List<TodoExemption>,
+  repoPath: String
+): List<Pair<String, Int>> {
+  return todoExemptionList.flatMap { exemption ->
+    exemption.getLineNumberList().mapNotNull { exemptedLineNumber ->
+      if (
+        !todos.any {
+          it.filePath.removePrefix(repoPath) == exemption.exemptedFilePath &&
+            it.lineNumber == exemptedLineNumber
+        }
```

Comment:
Suggest moving this to a val to help make this bit of a code a bit more readable.

Reply:
Done.